mino/httpclient/mod.go

```diff
@@ -0,0 +1,140 @@
+package httpclient
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"strings"
+	"time"
+)
+
+type key int
+
+const (
+	requestIDKey key = 0
+)
+
+// Httpclient defines the primitives to implement an http client that handles
+// client side requests
+type Httpclient interface {
+	// Start starts the http server. This call is assumed to be blocking
+	Start()
+
+	// RegisterHandler registers a new handler
+	RegisterHandler(path string, handler func(http.ResponseWriter, *http.Request)) error
+}
+
+// NewClient creates a new http client
+func NewClient(listenAddr string) *Client {
+	logger := log.New(os.Stdout, "httpclient :", log.LstdFlags)
+
+	nextRequestID := func() string {
+		return fmt.Sprintf("%d", time.Now().UnixNano())
+	}
+
+	mux := http.NewServeMux()
+
+	return &Client{
+		mux: mux,
+		server: &http.Server{
+			Addr:    listenAddr,
+			Handler: tracing(nextRequestID)(logging(logger)(mux)),
+		},
+		logger:     logger,
+		listenAddr: listenAddr,
+	}
+}
+
+// Client defines an http client
+//
+// - implements Httpclient
+type Client struct {
+	mux        *http.ServeMux
+	server     *http.Server
+	logger     *log.Logger
+	listenAddr string
+}
+
+// Start implements Httpclient
+func (s Client) Start() {
+	s.logger.Println("Client server is starting...")
+
+	done := make(chan bool)
+	quit := make(chan os.Signal, 1)
+	signal.Notify(quit, os.Interrupt)
+
+	go func() {
+		<-quit
+		s.logger.Println("Server is shutting down...")
```

Comment:
You can make it simpler (and easier to test) by removing the signal catching and having _Start_ and _Stop_ functions. When the daemon is stopped, it should be possible to call _Stop_ but it is yet to be implemented.

Reply:
Also rename Start to Listen and the function can be blocking.